Turborepo導入

Hiroto-Shida opened the pull request:
Turborepoを導入し、ディレクトリ構成・パッケージ管理・使用コマンド等が大きく変わりました

## やったこと

- Turborepo のディレクトリ構成にし、必要なファイル等を増やした
  - [Turborepo公式](https://turborepo.com/)、[Zenn](https://zenn.dev/hayato94087/articles/d2956e662202a7)
  - ざっくり以下構成
```
/
├─ apps
│     ├─ storybook（----- storybook -----）
│     │  ├─ package.json
│     │  └─ ...
│     └─ web（----- Next.js フロント -----）
│        ├─ src
│        │  └─ app
│        │     └─ ...
│        └─ package.json
├─ packages
│     ├─ tailwind-config（----- Tailwind の設定 -----）
│     │  └─ package.json
│     ├─ typescript-config（----- Typescript の設定 -----）
│     │  └─ package.json
│     └─ ui（----- 共通UI(components) -----）
│        └─ package.json
├─ biome.json
├─ package.json
├─ pnpm-lock.yaml
├─ tsconfig.json
└─ turbo.json
```

- パッケージ管理を npm から pnpm にした(turborepo の推奨がpnpmだったため)
- GHA workflow の修正
  - pnpm を使う都合上ちょっと変更が入った
  - 最初pnpmの公式GHAを使おうとしたがあまりメリットがないっぽく、corepackを使う方法にした（[参考](https://askai.glarity.app/ja/search/GitHub-Actions-%E3%81%A7-pnpm-%E3%82%92%E4%BD%BF%E3%81%86%E6%96%B9%E6%B3%95)）
- react three の再install
- tailwind周りの整理（turborepoを使う上でいろいろ整理が必要になったので再構築）
  - `packages/tailwind-config` にてtailwindやpostcss等の設定を一元管理
  - `packages/ui`や`apps/web`などのtailwindを使いたいディレクトリでは、以下行うことで使えるようになる
    - dependenciesに`tailwindcss`の追加
    - `postcss.config.mjs`の用意
    - cssファイルを用意し、中でtailwindcssや共通スタイルのimport（`@source`はpackages/uiのstyleを持ってくるために必要だった）
  - いろんな階層構造になったことで拡張機能のintellisenseが効かなくなったので、settings.jsonに`tailwindCSS.experimental.configFile`を追加して対処（[参考](https://stackoverflow.com/questions/79394239/vscode-tailwind-css-intellisense-autocompletion-and-suggestions-not-working-afte)）

## 追加必要な環境構築手順

- pnpmコマンドを使えるようにするまで
```
volta uninstall pnpm（必要であれば。すでに入ってると上手く合わないかも??）
volta install corepack
corepack enable
pnpm -v（ルートの package.json "packageManager"と同じversionになればOK）
```

- （任意）turbo のグローバルインストール
  - [公式Doc](https://turborepo.com/docs/getting-started/installation#installing-turbo)では入れとけば便利だよって言ってる
  - turboコマンドが使えるとルートの`turbo.json`のコマンドがそのまま使える
    - `turbo dev`（=`pnpm run dev`）、
    - `turbo dev`（=`pnpm run dev`）、など
```
pnpm add turbo --global
```

- 初回パッケージインストール
```
pnpm i --frozen-lockfile
```

- 起動（現状storybookとNext.jsが同時起動）
```
pnpm run dev　もしくは　turbo dev
```

## その他使用可能コマンド

- 新しいパッケージを入れるとき（`-D`でdevDependenciesに入れる）
```
pnpm add xxx
pnpm add -D xxx
```

- 特定のディレクトリ内(以下uiの例)に新しいパッケージを入れるとき
```
pnpm add xxx --filter ui
pnpm add -D ui
```

- その他はpackage.jsonを参照

## TODO

- テスト環境の整備
  - `apps/storybook`のdevDependenciesにvitestなどは残ってるが使える状態では無い
  - 設定ファイルもろもろ一旦消したので、最終的に`pnpm run test`とかで実行できる状態にできればよさそう?
- [turbo/gen](https://turborepo.com/docs/guides/generating-code) の導入
  - 本turborepoで同じく推奨されている雛形生成ツール
    - plopと同じようなものだが、turboで使うならこれがよさそう
    - [turbo/gen記事](https://zenn.dev/json_hardcoder/articles/1e8628dab0f31c)、[prop記事](https://zenn.dev/counterworks/articles/plop-template)

Hiroto-Shida committed:
一旦buildとdev起動ができるとこまで

Hiroto-Shida committed:
biomeの設定変更

Hiroto-Shida committed:
一括format(tabはやめた)

Hiroto-Shida committed:
buildのworkflowを修正

Hiroto-Shida committed:
corepackによるpnpmインストールに変えてみる

Hiroto-Shida committed:
storybookのpublicディレクトリが不要のため削除

Hiroto-Shida committed:
lint and format checkもpnpmで動くように修正

Hiroto-Shida committed:
biomeのGHAをやめるように修正

Hiroto-Shida committed:
rootでtypescriptを管理するように修正し、voltaで全体管理

Hiroto-Shida committed:
不要なtsconfig削除

Hiroto-Shida committed:
biomeのGHAを使うように元に戻す

Hiroto-Shida committed:
ためしにformat errorを発生させてみる

Hiroto-Shida committed:
format errorを直す

Hiroto-Shida committed:
typescript-configの不要なpublishse設定を削除

Hiroto-Shida committed:
react-threeの導入

Hiroto-Shida committed:
tailwind大整理

Hiroto-Shida committed:
gitignore統合

Hiroto-Shida committed:
storybookとnext.jsのビルドを分けるように変更

INSERT-io commented on apps/storybook/.storybook/main.ts:
```diff
@@ -0,0 +1,29 @@
+import type { StorybookConfig } from '@storybook/experimental-nextjs-vite';
+
+const config: StorybookConfig = {
+  stories: ['../../../packages/ui/src/**/*.stories.@(js|jsx|mjs|ts|tsx)'],
```
絶対パスで書いたりはできないですかね？
無理であれば大丈夫です

Hiroto-Shida replied:
できなくはなさそうですが、設定を色々追加したり、結構手を加えないと行けなさそうなのでやめます

INSERT-io commented on apps/storybook/.storybook/globals.css:
```diff
@@ -0,0 +1,4 @@
+@import "tailwindcss";
+@import "@repo/tailwind-config";
+
+@source "../../../packages/ui/src/";
```
これも絶対パスにできたりするのかなと気になったり

Hiroto-Shida replied:
できませんでしたので許して

INSERT-io commented on .vscode/settings.json:
```diff
@@ -1,19 +1,25 @@
+  "typescript.tsdk": "node_modules/typescript/lib",
+  "editor.formatOnSave": true,
+  "editor.tabSize": 2,
+  "tailwindCSS.experimental.configFile": {
+    "apps/storybook/.storybook/globals.css": "apps/storybook/.storybook/**",
+    "apps/web/src/app/globals.css": "apps/web/src/app/**",
+    "packages/ui/src/styles.css": "packages/ui/src/**"
+  }
```
調べても微妙に理解できなかったんですがこれって何の機能ですかね？

Hiroto-Shida replied:
PRコメントにも貼ってるけど、以下参考ですね。
https://stackoverflow.com/questions/79394239/vscode-tailwind-css-intellisense-autocompletion-and-suggestions-not-working-afte

tailwindの拡張機能のintellisenseがv4だと、この設定でパスを具体的に書いてあげないとちゃんと効かないっぽいです

INSERT-io reviewed (commented):
おおよそ問題ないと思いますが数点コメントしてます！
あと、READMEの更新があると嬉しいです！(起動コマンドも変わったり増えたりしたと思うので)

Hiroto-Shida commented:
@INSERT-io なんも直してないけどコメント返しました！

Hiroto-Shida committed:
:memo: READMEを更新

Hiroto-Shida commented:
@INSERT-io READMEも更新しました！

Hiroto-Shida commented:
GHAのbiomeがミスってるので対処PRを別ではやします

Hiroto-Shida committed:
:up: biomeを2.2.3にupdate

Hiroto-Shida committed:
:gear: Your configuration file is up to date.
configuration ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  i Migration results:

  - /Users/shidahiroto/projects/study/rubikstation-new/biome.json: no migration needed.
  を実行し、v2系のbiome.jsonに更新

Hiroto-Shida committed:
:+1: biomeのincludesを修正

Hiroto-Shida committed:
:gear: check_lintのbiomeにversion指定

INSERT-io reviewed (approved):
よさそう

Hiroto-Shida committed:
:up: biomeを2.2.4へupdate

Hiroto-Shida committed:
:gear: biome.jsonのschemaのversionも2.2.4に修正

Hiroto-Shida committed:
:gear: biome.jsonのincludesのglobals.cssを全cssに修正

Hiroto-Shida committed:
:shirt: format-and-lint実行

Hiroto-Shida committed:
Merge pull request #12 from Hiroto-Shida/feature/update-biome-2.2.4

biomeを2.2.4へupdate

INSERT-io reviewed (approved):
よさそう
